File: EXE1_Java/Q9.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

public class Q9 {
    static class Order {
        String name;
        PriorityQueue<Integer> queue = new PriorityQueue<>(Comparator.reverseOrder());  // this is max heap
        int totalCount = 0;  // waiting for remaining food count
        int waitedMinutes = 0;

        Order(String s) {
            String[] s2 = s.split("#");
            name = s2[0];
            for (String s3 : s2[1].split("%")) {
                int count = Integer.parseInt(s3.split(":")[1]);
                queue.add(count);
                totalCount += count;
            }
        }

        void doWait() {
            if (totalCount == 0) return;
            waitedMinutes += 5;
        }
    }

    static class Cook {
        Order cookingFor;
        int remainingCount = 0;

        void acquireFood(List<Order> orders) {
            if (remainingCount != 0) return;
            for (Order o : orders) {
                if (o.queue.isEmpty()) continue;
                cookingFor = o;
                remainingCount = o.queue.poll();
                return;
            }
        }

        void cook() {
            if (remainingCount == 0) return;
            remainingCount--;
            cookingFor.totalCount--;
        }
    }

    static boolean needWaiting(List<Order> orders) {
        for (Order o : orders) {
            if (o.totalCount > 0) return true;
        }
        return false;
    }

    static void Q9(List<String> sb) {
        String s = String.join("", sb).replaceAll(" ", "");
        List<Order> orders = new ArrayList<>();
        for (String substr : s.split(";")) {
            orders.add(new Order(substr));
        }
        // process order now
        List<Cook> cooks = new ArrayList<>();
        for (int i = 0; i < 3; i++) cooks.add(new Cook());
        while (needWaiting(orders)) {
            for (Cook c : cooks) c.acquireFood(orders);

            for (Order o : orders) o.doWait();
            for (Cook c : cooks) c.cook();
        }
        for (Order o : orders) {
            int hr = 10;
            int min = 0;
            min += o.waitedMinutes;
            hr += min / 60;
            min %= 60;
            System.out.printf("%s can collect food at %02d:%02d\n", o.name, hr, min);
        }
    }

    public static void main(String[] args) {
        BufferedReader sr = new BufferedReader(new InputStreamReader(System.in));
        String s = null;
        do {
            try {
                List<String> sb = new ArrayList<>();
                while ((s = sr.readLine()) != null) {
                    if (s.length() != 0) sb.add(s);
                    else if (sb.size() != 0) break;
                }
                Q9(sb);
            } catch (Exception e) {
                e.printStackTrace();
            }
        } while (s != null);
    }
}

File: EXE1_Python/Q9.py
#!/usr/bin/env python3

import sys
import traceback
import heapq

class MaxHeap:
    """A object-oriented max heap implementation by wrapping python heapq module
    """
    def __init__(self):
        self.heap = []

    def push(self, num):
        heapq.heappush(self.heap, -num)  # heapq is min heap

    def pop(self):
        return -heapq.heappop(self.heap)

    def peek(self):
        return -self.heap[0]

    def __len__(self):  # define __len__ for __bool__ which is for emptiness checking
        return len(self.heap)

class Order:
    def __init__(self, s):
        self.name = None
        self.queue = MaxHeap()
        self.totalCount = 0
        self.waitedMinutes = 0

        s2 = s.split('#')
        self.name = s2[0]
        for s3 in s2[1].split('%'):
            count = int(s3.split(':')[1])
            self.queue.push(count)
            self.totalCount += count

    def doWait(self):
        if self.totalCount == 0:
            return
        self.waitedMinutes += 5

class Cook:
    def __init__(self):
        self.cookingFor = None
        self.remainingCount = 0

    def acquireFood(self, orders):
        if self.remainingCount != 0:
            return
        for o in orders:
            if o.queue:  # if queue not empty
                self.cookingFor = o
                self.remainingCount = o.queue.pop()
                return

    def cook(self):
        if self.remainingCount != 0:
            self.remainingCount -= 1
            self.cookingFor.totalCount -= 1

def needWaiting(orders):
    for o in orders:
        if o.totalCount > 0:
            return True
    return False

def Q9(sb):
    s = ''.join(sb).replace(' ', '')
    orders = [Order(substr) for substr in s.split(';')]
    # process order now
    cooks = [Cook() for _ in range(3)]
    qq = 0
    while needWaiting(orders):
        for c in cooks:
            c.acquireFood(orders)

        for o in orders:
            o.doWait()
        for c in cooks:
            c.cook()

        qq += 1
        if qq > 1000:
            raise RuntimeError('error: never ends...')

    for o in orders:
        hr = 10
        min = 0
        min += o.waitedMinutes
        hr += min // 60  # int divison
        min %= 60
        print('%s can collect food at %02d:%02d' % (o.name, hr, min))


if __name__ == '__main__':
    """Verify test cases easily
    1. run the program
    2. paste test case to stdin
    3. press double Enter
    4. continue pasting next test case
    """
    do_next = True
    while do_next:
        try:
            do_next = False  # do_next won't set if encounter EOF
            sb = []
            for line in sys.stdin:  # s = data + '\n'
                line = line.rstrip('\r\n')
                if line:  # check str not empty
                    sb.append(line)
                elif sb:  # check list not empty
                    do_next = True
                    break
            Q9(sb)
        except Exception:
            traceback.print_exc()
